ibis/expr/decompile.py

```diff
@@ -0,0 +1,409 @@
+    right = _maybe_add_parens(op.right, right)
+    return f"{left} {operator} {right}"
+
+
+class Varnames:
+    # never assign columns and primitive datatypes to variables
+    ignored = (ops.TableColumn, dt.Primitive, dt.Variadic, dt.Temporal)
+    aliases = {
+        ops.ScalarParameter: "param",
+        ops.Literal: "l",
+        ops.TableNode: "t",
+        ops.Aggregation: "agg",
+        ops.Selection: "proj",
+        ops.NodeList: "lst",
+    }
+
+    def __init__(self):
+        self.counters = collections.defaultdict(itertools.count)
+
+    def generate(self, node):
+        if isinstance(node, self.ignored):
+            return None
+
+        if isinstance(node, ops.TableNode) and isinstance(node, ops.Named):
+            name = node.name
+        else:
+            klass = node.__class__
+            name = self.aliases.get(klass, klass.__name__.lower())
+
+        # increment repeated type names: table, table1, table2, ...
+        num = next(self.counters[name]) or ''
+        return f"{name}{num}"
+
+
+@experimental
+def decompile(node, render_import=True, assign_result=True, format=False):
+    """Decompile an ibis expression into Python source code.
+
+    Parameters
+    ----------
+    node : ir.Expr
+    render_import : bool, default True
+        Whether to add 'import ibis' import statement.
+    assign_result : bool, default True
+        Whether to assign the result to a variable called 'result'.
+    format : bool, default False
+        Whether to format the generated code using black code formatter.
+
+    Returns
+    -------
+    code : str
+    """
+    if isinstance(node, ir.Expr):
+        node = node.op()
+
+    out = io.StringIO()
+    varnames = Varnames()
+    dependents = Graph(node).invert()
+
+    def fn(node, *args, **kwargs):
+        # TODO(kszucs): factor the logic out to a context object which can be
+        # passed to the translate function
+        result = translate(node, *args, **kwargs)
+        if not result:
+            return None
+
+        n_dependents = len(dependents[node])
+
+        if n_dependents > 1 or isinstance(node, ops.ScalarParameter):
+            # multiple dependents, so we need to assign a variable
+            if var := varnames.generate(node):
+                out.write(f"{var} = {result}\n")
+                return var
+            else:
+                return result
+        elif n_dependents == 1:
+            # only one dependent, so we can inline the expression
+            return result
+        else:
+            # no dependents, so we deal with the root node
+            if assign_result:
+                out.write(f"\nresult = {result}\n")
+            else:
+                out.write(str(result))
+            return result
+
+    node.map(fn)
+
+    result = out.getvalue()
+    if render_import:
+        result = f"import ibis\n\n\n{result}"
+    if format:
+        import black
+
+        result = black.format_str(result, mode=black.FileMode())
```

Comment:
Will this pickup configuration using black's default `pyproject.toml` search procedure?

Reply:
Great question, I have not checked it.